Couldn't find DT_RPATH tag

When running a staticx-bundled program, we sometimes see this error:
```
myprog: Couldn't find DT_RPATH tag
```

I tried running this myself with the `ubuntu:xenial` docker image, and I hit yet another weird issue:
```
# cd test/rpath-origin
# scons
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
scons: building associated VariantDir targets: build
gcc -o build/app.o -c -Wall -Werror app.c
gcc -o build/libfoo.os -c -Wall -Werror -fPIC libfoo.c
gcc -o build/libbar.os -c -Wall -Werror -fPIC libbar.c
gcc -o build/libbar.so -z origin -shared -Wl,-rpath=\$ORIGIN build/libbar.os -Ldist/lib
Install file: "build/libbar.so" as "dist/lib/libbar.so"
gcc -o build/libfoo.so -z origin -shared -Wl,-rpath=\$ORIGIN build/libfoo.os -Ldist/lib -lbar
Install file: "build/libfoo.so" as "dist/lib/libfoo.so"
gcc -o build/app -z origin -Wl,-rpath=\$ORIGIN/../lib build/app.o -Ldist/lib -lfoo
/usr/bin/ld: warning: libbar.so, needed by dist/lib/libfoo.so, not found (try using -rpath or -rpath-link)
dist/lib/libfoo.so: undefined reference to `bar_nop'
collect2: error: ld returned 1 exit status
scons: *** [build/app] Error 1
scons: building terminated because of errors.
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,12 @@
 All notable changes to this project will be documented in this file.
 This project adheres to [Semantic Versioning](http://semver.org/).
 
+## [Unreleased]
+### Fixed
+- Worked around patchelf bug which caused `Couldn't find DT_RPATH tag`
+  error at runtime ([#175])
+
+
 ## [0.12.1] - 2021-02-06
 ### Fixed
 - Fixed bug causing libnssfix to be built incorrectly under
@@ -218,3 +224,4 @@ Initial release
 [#154]: https://github.com/JonathonReinhart/staticx/pull/154
 [#157]: https://github.com/JonathonReinhart/staticx/pull/157
 [#168]: https://github.com/JonathonReinhart/staticx/pull/168
+[#175]: https://github.com/JonathonReinhart/staticx/pull/175

--- a/staticx/elf.py
+++ b/staticx/elf.py
@@ -194,6 +194,12 @@ def patch_elf(path, interpreter=None, rpath=None, force_rpath=False, no_default_
     if rpath:
         args += ['--set-rpath', rpath]
     if force_rpath:
+        # There is a bug in patchelf that requires --remove-rpath to be used
+        # first before --force-rpath is effective.
+        # https://github.com/NixOS/patchelf/issues/94
+        # This was fixed in v0.11 but that's newer than Debian Buster.
+        tool_patchelf.run_check('--remove-rpath', path)
+
         args.append('--force-rpath')
     if add_needed:
         for lib in coerce_sequence(add_needed):

--- a/test/rpath-origin/SConscript
+++ b/test/rpath-origin/SConscript
@@ -1,5 +1,7 @@
 Import('env')
 
+# app --deps--> libfoo.so --deps--> libbar.so
+
 lib = env.SharedLibrary(
     target = 'bar',
     source = ['libbar.c'],

--- a/test/rpath-origin/SConstruct
+++ b/test/rpath-origin/SConstruct
@@ -1,11 +1,14 @@
 env = Environment(
-    BUILD_DIR = '#build',
-    DIST_DIR = '#dist',
+    BUILD_DIR = Dir('#build'),
+    DIST_DIR = Dir('#dist'),
     LIB_DIR = '$DIST_DIR/lib',
     BIN_DIR = '$DIST_DIR/bin',
 
     # https://github.com/SCons/scons/wiki/UsingOrigin
-    LINKFLAGS = ['-z', 'origin'],
+    LINKFLAGS = [
+        '-z', 'origin',
+        '-Wl,-rpath-link', '$LIB_DIR',
+    ],
 
     # Programs look for shared libs in ../lib
     PROG_RPATH = Literal('\\$$ORIGIN/../lib'),
@@ -16,6 +19,9 @@ env = Environment(
     CCFLAGS = ['-Wall', '-Werror'],
     LIBPATH = ['$LIB_DIR'],
 )
+
+# gcc (ld) requires that -rpath-link dirs be present, but they may not (yet) exist
+env.Execute(Mkdir('$LIB_DIR'))
 
 env.SConscript(
     'SConscript',

